pkg/operator/ceph/object/objectstore.go

```diff
@@ -56,129 +62,448 @@ type idType struct {
+	}
+	logger.Debugf("found keys secret for realm %v", realmName)
+
+	accessKey, err := DecodeSecret(realmSecret, AccessKeyName)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to decode realm %q access key", realmName)
+	}
+	secretKey, err := DecodeSecret(realmSecret, SecretKeyName)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to decode realm %q access key", realmName)
+	}
+	logger.Debugf("decoded keys for realm %v", realmName)
+
+	accessKeyArg := fmt.Sprintf("--access-key=%s", accessKey)
+	secretKeyArg := fmt.Sprintf("--secret-key=%s", secretKey)
+
+	return accessKeyArg, secretKeyArg, nil
+}
+
+func getZoneEndpoints(objContext *Context, serviceEndpoint string) ([]string, error) {
+	logger.Debugf("getting current endpoints for zone %v", objContext.Zone)
+	realmArg := fmt.Sprintf("--rgw-realm=%s", objContext.Realm)
+	zoneGroupArg := fmt.Sprintf("--rgw-zonegroup=%s", objContext.ZoneGroup)
+
+	zoneGroupOutput, err := RunAdminCommandNoMultisite(objContext, "zonegroup", "get", realmArg, zoneGroupArg)
+	if err != nil {
+		return []string{}, errors.Wrap(err, "failed to get rgw zone group")
+	}
+	zoneGroupJson, err := DecodeZoneGroupConfig(zoneGroupOutput)
+	if err != nil {
+		return []string{}, errors.Wrap(err, "failed to parse zones list")
+	}
+
+	zoneEndpointsList := []string{}
+	for _, zone := range zoneGroupJson.Zones {
+		if zone.Name == objContext.Zone {
+			for _, endpoint := range zone.Endpoints {
+				// in case object-store operator code is rereconciled, zone modify could get run again with serviceEndpoint added again
+				if endpoint != serviceEndpoint {
+					zoneEndpointsList = append(zoneEndpointsList, endpoint)
+				}
+			}
+			break
+		}
+	}
+
+	return zoneEndpointsList, nil
+}
+
+func createMultisite(objContext *Context, endpointArg string) error {
+	logger.Debugf("creating realm, zone group, zone for object-store %v", objContext.Name)
+
+	realmArg := fmt.Sprintf("--rgw-realm=%s", objContext.Name)
+	zoneGroupArg := fmt.Sprintf("--rgw-zonegroup=%s", objContext.Name)
+
+	updatePeriod := false
+	// create the realm if it doesn't exist yet
+	output, err := RunAdminCommandNoMultisite(objContext, "realm", "get", realmArg)
+	if err != nil {
 		updatePeriod = true
-		_, err := RunAdminCommandNoRealm(context, "zonegroup", "modify", realmArg, zoneGroupArg, endpointArg)
+		output, err = RunAdminCommandNoMultisite(objContext, "realm", "create", realmArg)
 		if err != nil {
-			return errors.Wrapf(err, "failed to add object store %q in rgw zone group %q", context.Name, zoneGroupName)
+			return errors.Wrapf(err, "failed to create rgw realm %q for reason: %q", objContext.Name, output)
 		}
-		_, err = RunAdminCommandNoRealm(context, "zone", "modify", realmArg, zoneGroupArg, zoneArg, endpointArg)
+	}
+	logger.Debugf("created realm %v", objContext.Name)
+
+	// create the zonegroup if it doesn't exist yet
+	output, err = RunAdminCommandNoMultisite(objContext, "zonegroup", "get", realmArg, zoneGroupArg)
+	if err != nil {
```

Comment:
Can we detect a more specific error code for it not existing? Otherwise, any error could result in attempting to create the zonegroup.